Operational Semantics

SethPoulsen opened the pull request:
Beginning implementation of operational semantics. 

Still need to implement `return` and the alternative `cutpoint` semantics that will make operational equivalent to verification semantics

nishantjr committed:
build: Add test infrastructure for operational semantics; stop using lit

nishantjr committed:
boogie: Split configuration for types/unique information

nishantjr committed:
split out two new modules RUNTIME and HELPERS

nishantjr committed:
Formatting

nishantjr committed:
boogie: Add alternate rule for `#start` for operational and verification semantics

nishantjr committed:
bin/kboogie: fix default definition namem

nishantjr committed:
bin/kboogie: Fix wrong definition being used to unparse

nishantjr committed:
bin/diff-kboogie: Avoid printing empty line on success

SethPoulsen committed:
Able to run operational semantics on program with just a main

SethPoulsen committed:
Fix typo in code block label

SethPoulsen committed:
Disable most verification tests for quicker dev on operational (will revert before merge)

SethPoulsen committed:
Added expected files for operational tests. Some simple tests pass.

SethPoulsen committed:
Enabled a few more passing tests

SethPoulsen committed:
Added test cases from the Boogaloo project

source: https://bitbucket.org/nadiapolikarpova/boogaloo/src/default/examples/rv2013/

SethPoulsen committed:
Added .expect files for Boogaloo tests

SethPoulsen committed:
Add lit test commands to Boogaloo test files

SethPoulsen committed:
Added simple procedure call example

SethPoulsen committed:
Disable warnings from Kompile

SethPoulsen committed:
boogie: change syntax definition to allow eager eval of function application in call statement

SethPoulsen committed:
boogie: stack to save important info when calling a procedure

SethPoulsen committed:
Added simple example showing when operational and verification semantics differ

SethPoulsen committed:
Revert "Disable most verification tests for quicker dev on operational (will revert before merge)"

This reverts commit c5e9e466152f4c50678c4aef6c1dbe01bed3c0b3.

SethPoulsen committed:
Disable some of the newly added tests that don't pass yet

SethPoulsen commented on test/failing.verification:
```diff
@@ -0,0 +1,93 @@
+# Not sure why they're not passing, looks like a whitespace issue in the diff?
+test/operational/control-flow/while-no-invar.bpl
+test/operational/control-flow/square-proc.bpl
```
@nishantjr These test _should_ pass but don't. It looks like the only diff by `diff-kboogie` is whitespace. I'm not sure what's going on there could you have a look?

nishantjr replied:
Looks like a bug in substitution actually.

SethPoulsen replied:
You mean the substitution being done by `diff-kboogie`? or one of the substitution rules in the `k` definition?

nishantjr replied:
I thought it was the one we defined in the boogie repository. But turns out, I wasn't evaluating the arguments of the function to values before applying them. Pushed a fix

nishantjr replied:
The "fix" breaks things more, since functions with symbolic aren't evaluated. :shrug:

SethPoulsen replied:
All right. Well, we can just revert the "fix" so that we can merge this with `master` with all tests passing, then worry about it later.

nishantjr committed:
runtime: call: Evaluate procedure application arguments before calling

```
procedure p(int: x) returns(int: y)... ensures y = x * x;
call w := p(w);
```

evaluated to `w = w*w` here, instead of `w = old(w) * old(w)`

nishantjr committed:
runtime: call: Add TODO re missing assertion

nishantjr committed:
Revert "runtime: call: Evaluate procedure application arguments before calling"

This reverts commit acc2fc33e3d852a6f16d7908f413131c36e2737a.

nishantjr commented:
Subsumed into #7
